src/sys/fanotify.rs

```diff
@@ -0,0 +1,365 @@
+        /// Synchronized I/O data integrity completion.
+        O_DSYNC as libc::c_uint;
+        /// No file last access time update.
+        O_NOATIME as libc::c_uint;
+        /// Nonblocking mode for the file descriptor.
+        O_NONBLOCK as libc::c_uint;
+        /// Synchronized I/O file integrity completion.
+        O_SYNC as libc::c_uint;
+    }
+}
+
+libc_bitflags! {
+    /// Configuration options for [`fanotify_mark`](fn.fanotify_mark.html).
+    pub struct MarkFlags: libc::c_uint {
+        /// Add the events to the marks.
+        FAN_MARK_ADD;
+        /// Remove the events to the marks.
+        FAN_MARK_REMOVE;
+        /// Don't follow symlinks, mark them.
+        FAN_MARK_DONT_FOLLOW;
+        /// Raise an error if filesystem to be marked is not a directory.
+        FAN_MARK_ONLYDIR;
+        /// Events added to or removed from the marks.
+        FAN_MARK_IGNORED_MASK;
+        /// Ignore mask shall survive modify events.
+        FAN_MARK_IGNORED_SURV_MODIFY;
+        /// Remove all marks.
+        FAN_MARK_FLUSH;
+        /// Do not pin inode object in the inode cache. Since Linux 5.19.
+        FAN_MARK_EVICTABLE;
+        /// Events added to or removed from the marks. Since Linux 6.0.
+        FAN_MARK_IGNORE;
+
+        /// Default flag.
+        FAN_MARK_INODE;
+        /// Mark the mount specified by pathname.
+        FAN_MARK_MOUNT;
+        /// Mark the filesystem specified by pathname. Since Linux 4.20.
+        FAN_MARK_FILESYSTEM;
+
+        /// Combination of `FAN_MARK_IGNORE` and `FAN_MARK_IGNORED_SURV_MODIFY`.
+        FAN_MARK_IGNORE_SURV;
+    }
+}
+
+/// Compile version number of fanotify API.
+pub const FANOTIFY_METADATA_VERSION: u8 = libc::FANOTIFY_METADATA_VERSION;
+
+#[derive(Debug)]
+/// Abstract over `libc::fanotify_event_metadata`, which represents an event
+/// received via `Fanotify::read_events`.
+pub struct FanotifyEvent {
+    /// Version number for the structure. It must be compared to
+    /// `FANOTIFY_METADATA_VERSION` to verify compile version and runtime
+    /// version does match. It can be done with the
+    /// `FanotifyEvent::has_compile_version` method.
+    pub version: u8,
+    /// Mask flags of the events.
+    pub mask: MaskFlags,
+    /// The file descriptor of the event. If the value is `None` when reading
+    /// from the fanotify group, this event is to notify that a group queue
+    /// overflow occured.
+    pub fd: Option<OwnedFd>,
+    /// PID of the process that caused the event. TID in case flag
+    /// `FAN_REPORT_TID` was set at group initialization.
+    pub pid: i32,
+}
+
+impl FanotifyEvent {
+    /// Checks that compile fanotify API version is equal to the version of the
+    /// event.
+    pub fn has_compile_version(&self) -> bool {
+        self.version == FANOTIFY_METADATA_VERSION
+    }
+}
+
+#[derive(Debug)]
+/// Abstraction over the structure to be sent to allow or deny a given event.
+pub struct FanotifyResponse<'e> {
```

Comment:
And this one:

```rs
#[derive(Debug)]
#[repr(transparent)]
pub struct FanotifyResponse<'a> {
    inner: libc::fanotify_response,
    _borrowed_fd: PhantomData<BorrowedFd<'a>>,
}

impl<'a> FanotifyResponse<'a> {
    /// Create a new response.
    pub fn new(fd: BorrowedFd<'a>, response: Response) -> Self {
        let response = match response {
            Response::Allow => libc::FAN_ALLOW,
            Response::Deny => libc::FAN_DENY,
        };

        Self {
            inner: libc::fanotify_response {
                fd: fd.as_raw_fd(),
                response,
            },
            _borrowed_fd: PhantomData,
        }
    }
}
```

Reply:
Nice! I also declared `FAN_ALLOW` and `FAN_DENY` directly with `libc_bitflags`. In case you're wondering why not an enum instead, this is because another two values (`FAN_AUDIT` and `FAN_INFO`) will be added later to send additional data with the response, and they are declared in the C header as bitmask on allow/deny values.